### ISSUE TITLE
aws s3 cp doesn't work with prefix

### ISSUE THREAD

**Issue by fabiocesari** [aws/aws-cli#1454]
From a bucket containing millions of files, I want do download a few thousand based on their prefix. 

Note: using --exclude "*" --include "myprefix" works, but is impractical since it lists all the files and applies the filter afterwards.

The problem is aws s3 cp adds a trailing "/" after the prefix I specify. In the example below you can see that logs/2015-08-17 becomes logs/2015-08-17/ 
This additional "/" leads to an empty result set, since the keys in the bucket have the form logs/YYYY-MM-DD-HH-MM-SS-RANDOM).

> aws s3 cp s3://mybucket/logs/2015-08-17 . --recursive --dryrun --debug

2015-08-17 21:48:32,026 - MainThread - awscli.clidriver - DEBUG - CLI version: aws-cli/1.7.27 Python/2.7.6 Linux/3.14.13-c9, botocore version: 0.108.0
2015-08-17 21:48:32,026 - MainThread - awscli.clidriver - DEBUG - Arguments entered to CLI: ['s3', 'cp', 's3://mybucket/logs/2015-08-17', '.', '--recursive', '--dryrun', '--debug']
....
2015-08-17 21:48:32,088 - MainThread - botocore.endpoint - DEBUG - Making request for <botocore.model.OperationModel object at 0x7ff9d3ec3e50> (verify_ssl=True) with params: {'query_string': {u'prefix': u'logs/2015-08-17/', u'encoding-type': 'url'}, 'headers': {}, 'url_path': u'/mybucket', 'body': '', 'method': u'GET'}

**Comment by rayluo** [comment]
The current CLI "aws s3 cp" behavior (coincidentally?) complies with the "cp" behavior. We will need to think whether or how to support the use case you mentioned.

**Comment by geota** [comment]
+1 This issue bit me today.

What is the reason/use case for adding a trailing slash? 

I would like to filter on a partial prefix that does not correspond to a "directory". I was able to work around this by using bash to filter the files and piping it back to aws cp.

**Comment by kyleknap** [comment]
@fabiocesari @geota 

We do read the exclude/include rules literally, and do not add the '/' to the end. To do what you want to do, you need to add a ``*`` to the end of the partial prefix, which means match all characters after. So a combination like ``--exclude '*' --include 'myprefix*'`` should work.

Let me know if you have any more questions.

**Comment by fabiocesari** [comment]
@kyleknap please read my previous post. The trailing slash is clearly added when the python code places calls to the S3 API (see the debug transcript), and using the wildcards is impractical because they are applied after all files are listed, which results in thousands of unnecessary calls.

**Comment by emmekappa** [comment]
+1

**Comment by JordonPhillips** [comment]
@fabiocesari I see what you're saying. It looks like the `--recursive` is the heart of the issue. We're assuming that when the flag is set, the s3 path is intended to be a directory. As you've demonstrated, this is not always the intent.

**Comment by cce** [comment]
+1

**Comment by mtngld** [comment]
+1, Encountered this issue today

**Comment by bnferguson** [comment]
Running into this today as well. Looks like it affects `s3 sync` as well. `s3 ls` seems to be working as intended.

**Comment by JamesBromberger** [comment]
+1: Wanted to download a set of S3 log files that are date prefixed, not separated by a '/' path. Had to fetch all 5000+ files instead of just the ~20 matching my requirements (logging prefix set to s3://mybucket/S3logs/prefix/, so files are in s3://mybucket/S3logs/prefix/2015-11-30-03-28-33-XXXXXXXXX). aws-cli/1.7.29

**Comment by ralph-tice** [comment]
I think instead of monkeying with --recursive in this context, you should concat the --include filter with the prefix at https://github.com/aws/aws-cli/blob/develop/awscli/customizations/s3/filegenerator.py#L317-L318 which I started a PR... #1707

If you could specify a different delimiter, that would also satisfy the main use case here, but it's hardcoded at https://github.com/aws/aws-cli/blob/develop/awscli/customizations/s3/subcommands.py#L465 and using boto's default everywhere else, I think -- the paginator code is really hard to follow.

**Comment by JamesBromberger** [comment]
Rough and ready: with S3 logging, you end up with one prefix with a very large number of objects matching (under the '/' separator). If you only want a small section, say perhaps matching a specific year-month-day-hour:
for F in `aws s3 ls s3://bucket/pathprefix/2016-04-05-11 | cut -c 32-`; do aws s3 cp s3://koondart-logs-prod/S3logs/prodmainpersistentblobstore-prodblobstore-1sa83kxh4c3bl/$F . ; done

**Comment by mathom** [comment]
I agree with both of ralph's ideas. Efficient downloading by prefix is a must have.

**Comment by ralph-tice** [comment]
How do I get #1707 reviewed/merged ?  Has anyone besides me tested my branch?

**Comment by alex88** [comment]
Just in case someone wants a script this worked for me:
```
aws s3 ls s3://bucket/fileprefix | cut -c 32- | while IFS= read -r line; do aws s3 cp s3://bucket/$line .; done
```

**Comment by mslinn** [comment]
Here is a script"
```
#!/bin/bash

BUCKET=www.whatevercom
DIR=log
PREFIX=$DIR/something.2016-09-30-16

aws s3 ls s3://$BUCKET/$PREFIX | \
  awk '{ print $4; }' | \
  while IFS= read -r line; do 
    CMD="aws s3 cp s3://$BUCKET/$DIR/$line ."
    #echo "$CMD" # uncomment for debugging
    $CMD
  done
```

**Comment by iainelder** [comment]
This workaround uses gnu parallel to run several downloads at the same time to make it faster to fetch all the files. It will probably break if you hit a certain number of matching keys, but I haven't test to find out where that limit is.

```
MATCHING_KEYS="$(
aws s3api list-objects-v2 \
--bucket "${BUCKET}" \
--prefix "${KEY_PREFIX}" |
jq --raw-output ".Contents[].Key"
)"

# Number of jobs arbitrarily chosen as a balance between speed and stability.
parallel --jobs 12 "aws s3 cp "s3://${BUCKET}/{}" ." ::: $MATCHING_KEYS
```

**Comment by andpol** [comment]
Using @kyleknap's suggested solution hasn't even started copying and it's been running for 10 minutes now. I'm assuming it's trying to list all my objects under the `logs/`prefix and then filter client-side:
```
aws s3 cp --recursive --exclude '*' --include '2016-11-16-17*'  s3://bucket/logs/ .
```

Best workaround I've seen so far is to abandon the AWS CLI tools completely:
```
s3cmd get 's3://bucket/logs/2016-11-16-17*' .
```

**Comment by ralph-tice** [comment]
@andpol did you try my fix in #1707 ?  i don't think alternative tools are a valid discussion point in this issue.

**Comment by andpol** [comment]
Yeah, that's fair. I'd love for the issue to be resolved correctly, but since we're already talking about shell scripts with awk and parallel.. ;)

FWIW, I just tried out your PR branch, executing `aws s3 cp --recursive --exclude '*' --include '2016-11-16-17*'  s3://bucket/logs/ .`

Worked perfectly! Hopefully someone on the AWS team can review your changes and get them merged.

**Comment by ralph-tice** [comment]
@kyleknap @JordonPhillips @jamesls can we get some help closing this out?

**Comment by bwjjohnson** [comment]
+1

**Comment by lababidi** [comment]
+1

**Comment by ncaadam** [comment]
+1

**Comment by metzlar** [comment]
+1

**Comment by seankovacs** [comment]
+1...moving 19TB without the ability to target specific keys on different machines is a nightmare. the include/exclude stuff still has to list through every page unlike a path wildcard found in the s3 ls command.

**Comment by shahbour** [comment]
+1

**Comment by bootrino** [comment]
Seems kinda crazy Ive not tried about 40 different aws s3 cp command variants and still can't find the incantation to copy the contents of a bucket folder to my local disk.  Puzzled as to why its so hard... shouldn't take hours to work out how to copy a few files.

**Comment by Plasma** [comment]
+1 just hit this; would have liked to just use ``aws s3 cp --recursive s3://bucket/2018-01-`` to grab logs for 2018-01-XX.

This partial match works in the S3 browser in AWS console, wish it worked for CLI.

**Comment by Gatisseja** [comment]
+1

**Comment by andychase** [comment]
@awstools Can we get an update on this?

**Comment by clandolfi** [comment]
+1
as andychase mentions above, an update would be great now that the footwork was done...

1,350,000 objects must be listed prior to returning the 20 we need? This basically renders s3 useless for our use case...

**Comment by mmarston** [comment]
+1

**Comment by hrez** [comment]
So, why can I ls but not cp a prefix? Is there some grand justification I missed in all those replies for the last 3 years?

**Comment by xmubeta** [comment]
+1， the same thing happens on 'aws s3 rm'.

**Comment by tdmalone** [comment]
> +1 just hit this; would have liked to just use `aws s3 cp --recursive s3://bucket/2018-01-` to grab logs for 2018-01-XX.

Here's a workaround for that case:
```aws s3 ls "s3://bucket/2018-01-" | awk '{print $4}' | xargs -I % aws s3 cp s3://bucket/% .```

**Comment by REslim30** [comment]
There have been a few PRs submitted for this issue, I think https://github.com/aws/aws-cli/pull/4423 is probably the closest so far but @ztane brings up a good point:

> I do not think this is correct though. I'd prefer another word instead of the --recursive for that behaviour, say --prefix, because --recursive has been associated with certain behaviour.
> yes but thousands of programs and assumptions would be broken. FWIW I want this feature, but only by different name - prefix, because that's what it is

We'd need to make sure it's backwards compatible, so putting it under a new flag would be a good solution. Prefix is a bit of an overloaded term (there's prefix in the S3 sense and prefix in the general sense), so maybe another term is required but I'm happy to revive that PR, introduce the new flag, put some tests in, and add docco.

**Comment by WesselAtWork** [comment]
Seems like most PR's are cancelled because it would break backwards compatibility in scripts. Or require further discussion to get down to a "better" approach.
I'll take a stab at the better approach:

Why not **"It was always a feature"**?
Introduce new default flags called something like `--delimiter=/` and `--auto-append-delimiter=true`
> `--delimiter` matches the interface in `aws s3api list-objects-v2`  

That way if people wanted to do things with recursive they could just do `--auto-append-delimiter=false` and if they wanted to play around with different delimiters you could do `--delimiter=-` or `--delimiter=@`.

You keep POSIX `cp` (don't know if that is really important because `ls` and `rm` aren't) and you get the power of s3 prefixes without breaking scripts.
Also doing the `aws s3 ls | xargs aws s3 cp` is not very fast when compared to a normal call to `aws s3 cp --recursive`.   
`aws s3 cp` only takes 2 args, the `SOURCE` and the `TARGET` and unlike posix `cp` can't take `cp SOURCE... TARGET` meaning you need to start N `aws` clis, that each need to do permission checks, that are each python programs and can't take advantage of pipelining optimizations; it ate my system alive when I tried `xargs -P0`.

You can then add these to the other interfaces `ls` and `rm` but instead make it default to `--auto-append-delimiter=false` then make it a ENV variable so everyone who wants to have a "consistent" way of interacting with the s3 command can do so by setting something like `AWS_S3_AUTO_APPEND_DELIMITER=FALSE` or  `AWS_S3_AUTO_APPEND_DELIMITER=TRUE` depending on their opinion on weather s3 prefixes are cool or not :P

I personally like this because it elegantly patches the "inconsistent" api in the other commands as "we like it that way; but if you don't, you can change it"
Of course if you see a way this can be borked please share.

> _As an asides something like `tar -T` would be awesome in `aws s3 cp` as it would allow us to create a lists of s3 files and then just simply put them into `aws s3` like: `aws s3 cp -T filelist.txt TARGET` or even cooler `aws s3 ls s3://whatever | aws s3 cp -T- .`_

**Comment by REslim30** [comment]
@WesselAtWork man I wish I did what you did and thought about this before jumping into an implementation. I was trying at adding an additional flag `--prefix` or `--partial-prefix` that would be an alternative to `--recursive` and work only with s3 to local or s3 to s3 copying. But tbh I don't think my approach is well thought out enough.

I feel like your approach is a lot more well thought out. I like the fact that it's mimicking an existing flag from `s3api` and how it provides the user and the ability to set the environment variables to get consistent behaviour.

Though I wonder if, from a user's point of view using `--recursive` and `--auto-append-delimiter=true` together would be a bit confusing. When a user runs this command:

```
aws s3 cp --recursive s3://aidand-misc/test .
```

They're trying to recurse into `test` as if it were a directory, they're not really thinking about delimiters. I would say that the --recursive flag should remain how it's currently behaving, as it's mimicking how cp --recursive works and it makes the analogy that prefixes are directories. I think approaches from here should be separate from --recursive, as we're now trying to treat s3 objects as `prefix + key` and so the --recursive analogy breaks down.

That being said, I'm not sure myself what the correct approach could be. Here's a little brainstorm:

Building on your delimiter idea, we could do something like this maybe:

```
aws s3 cp --delimiter=none 's3://aidand-misc/test*' .
```

Though this doesn't work as `*` is a valid s3 object character.


Maybe the whole analogy of `aws s3 cp` is breaking down here. And an analogy to a different command e.g., `aws s3 wget s3://aidand-misc/test*` could serve instead (maybe not wget but another command haha).

Maybe instead of having two positional arguments we could do something like:

```
aws s3 cp --prefix s3://aidand-misc/test --dest ~/foo
```

Which would specify `s3://aidand-misc/test` as being specifically a prefix. (Similar to how `aws s3api list-objects-v2 --prefix` works)

In any case, I think it'd be a good thing to discuss further to decide on a approach before coming up with an implementation.

**Comment by WesselAtWork** [comment]
@REslim30 

> Though I wonder if, from a user's point of view using --recursive and --auto-append-delimiter=true together would be a bit confusing.

That's why it is a default flag!
A new user will try to use the `aws s3 cp --recursive` the same way as normal without thinking about `--delimiter` or `--auto-append-delimiter`.
That will only happen when they want to be more specific.

---

> When a user runs this command: `aws s3 cp --recursive s3://aidand-misc/test .` They're trying to recurse into `test`

Consider these s3 keys:
```txt
test/something-1
test/something-2
test
test1
test2
test-something/1
test-something/2
```

---

Does the user want:
```txt
test/
├── something-1     <-
└── something-2     <-
test           <-?
test1
test2
test-something/
├── 1
└── 2
```
or
```txt
test/something-
├── 1
└── 2
test          <-?
test1
test2
test-
├── something/1     <-
└── something/2     <-
```
or even
```txt
test          <-?
test
├── /something-1     <-
├── /something-2     <-
├── 1                <-
├── 2                <-
├── -something/1     <-
└── -something/2     <-
```
---

Having the auto-append + delimiter it is 100% clear exactly what is going on and what the user wants.

---

== `aws s3 cp --recursive s3://aidand-misc/test`
== `aws s3 cp --recursive s3://aidand-misc/test/`
== `aws s3 cp --recursive --delimiter=/ --auto-append-delimiter s3://aidand-misc/test`
== `aws s3 cp --recursive --delimiter=/ s3://aidand-misc/test`
== `aws s3 cp --recursive --auto-append-delimiter s3://aidand-misc/test`
```txt
test/
├── something-1     <-
└── something-2     <-
```

or

== `aws s3 cp --recursive --auto-append-delimiter=false s3://aidand-misc/test-`
== `aws s3 cp --recursive --delimiter=- s3://aidand-misc/test`
== `aws s3 cp --recursive --auto-append-delimiter --delimiter=- s3://aidand-misc/test-`
```txt
test-
├── something/1     <-
└── something/2     <-
```

or

== `aws s3 cp --recursive --delimiter= s3://aidand-misc/test`
== `aws s3 cp --recursive --auto-append-delimiter --delimiter= s3://aidand-misc/test`
== `aws s3 cp --recursive --auto-append-delimiter=false s3://aidand-misc/test`
```txt
test          <-?
test
├── /something-1     <-
├── /something-2     <-
├── 1                <-
├── 2                <-
├── -something/1     <-
└── -something/2     <-
```
---

What happens with the `test` file in all of these?
We specified `--recursive` so it stands to reason you we to talk about it.
If we are working `S3 <-> S3` its fine.
But when working with a local filesystem you can't have a directory and a file with the same name.
It's not consistently bad, it can deal with file names like `test` and `test-something/1`.
What do?
Probably name it `file.<RANDOM>` if the error occurs as part of the download.
That or throw the OS "cannot create directory because file exists" exception and let the user deal with it.
¯\\_(ツ)_/¯

---

> it makes the analogy that prefixes are directories

s3 isn't a file system even though it [pretends](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-prefixes.html) to be. Delimiters are there to help you *organize*, they don't mean anything special and only help when you are listing.
We are just providing the "convenience" flag `--auto-append-delimiter` because people are used to working with filesystems.
Pretending is fine. As long as we have the ability to break out of it and use s3 to it's fullest potential if we understand and want take advantage of the fact that `s3 =/= fs`

> I think approaches from here should be separate from --recursive, as we're now trying to treat s3 objects as `prefix + key` and so the --recursive analogy breaks down.  
I don't quite understand this one. 
s3 objects are defined by their [key](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html).
Prefixes are partial keys that might or might not prefix other keys.

"Recursive" is pretty self explanatory to me and it is only weird because it clashes with existing terminology in the fs `cp`.

For that reason you could introduce another flag. 
Something called `--list` or `--expand` or `--iterate` or something else that means to consider the user input as a s3 prefix, not a key. (Basically pass it directly through to the `ls` interface)

Then you can do:  
`--recursive`  =  `--list --auto-append-delimiter=true --delimiter=/` - This is equivalent to normal `cp --recursive`
`--prefix`       =  `--list --auto-append-delimiter=false`                      - This is the s3 mode

---

> aws s3 cp --delimiter=none 's3://aidand-misc/test*' .

I like your `aws s3 cp s3://something*`  idea as another extension to the system. 
It allows you to specify the `--auto-append-delimiter=false` without needing to write it out, it would make these equivalent:
== `aws s3 cp --recursive --delimiter= s3://aidand-misc/test`  
== `aws s3 cp --recursive 's3://aidand-misc/test*'`  
== `aws s3 cp --recursive --auto-append-delimiter=false s3://aidand-misc/test`  

---

> Maybe we could actually make `--include` and `--exclude` work.

I think these are fine. They are a poor-man's `find -name` and `find ! -name`.
As long as it is clear that the way these works is by _filtering-at-the-client_ instead of the implied _filter-at-the-server_.
Someone else suggested upgrading the `--include "something*"` to internally use the much faster prefix mode instead of filtering the list...

Then again doing `aws ls my-prefix | grep 'REGEXP' | aws cp -T- .` will always be more powerful.
Maybe instead of trying to retroactively fix the broken `cp` interface it would be better (and easier) to just allow `cp` to take a convenient list from the user...

You can then also do:
`aws ls my-prefix | my-super-cool-filter-program.exe | aws cp -T- .`   
`my-super-cool-script.sh | aws cp -T- .`  
`my-program-that-outputs-files | aws cp -T- s3://somewhere/`  

---

Thinking on it now `--auto-append-delimiter` effectively means the same as `--append-delimiter` for future reference we should use that instead as it's shorter and cleaner.

---

> P.S. To me personally, both `mv` and `cp` should act like `ls` and we should educate users that they need to be specific. 
> As in `s3://bucket/something =/= s3://bucket/something/`.
> Instead of a `--recursive` flag there should be a `--file` flag, to differentiate when you are talking about a single file.
> To me it's way more convenient (and common) to refer to multiple files than a it is to want a single file.
> But that would be taking a **nuke** to the interface so no.

**Comment by REslim30** [comment]
Thanks for the comprehensive response, @WesselAtWork. You're going to think I'm a nitpick after reading this:

> s3 isn't a file system even though it [pretends](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-prefixes.html) to be. Delimiters are there to help you organize, they don't mean anything special and only help when you are listing.

For me it's this point, why `--recursive` doesn't make sense once you're using  `aws s3 cp --recursive --auto-append-delimeter=false s3://bucket/te .` to copy all of these files:

```
test/something-1
test/something-2
test
test1
test2
test-something/1
test-something/2
```

Since s3 isn't a filesystem, there are no directories to recurse into, rather you're just listing out all the objects in a bucket that start with the prefix `te`. In my mind, recursing implies a tree-like structure, whereas in this case, we're just listing out s3 objects. @ztane (who made a comment in a previous PR), I think feels the same way:

> I do not think this is correct though. I'd prefer another word instead of the --recursive for that behaviour, say --prefix, because --recursive has been associated with certain behaviour.
FWIW I want this fearure, but only by different name - prefix, because that's what it is

From the discussions above, there are a lot of people who think that `--recursive` is a good way to describe this behaviour - which I won't deny. And maybe I'm in the minority who think otherwise. But I would say there is a decent amount of people who would find the `--recursive` mental model not 100% applicable in this case. Which is to say the current behaviour of `aws s3 ls --recursive s3://aidand-misc/te` doesn't feel quite right imo.

I think your last solution would be super cool though:

> Then again doing aws ls my-prefix | grep 'REGEXP' | aws cp -T- . will always be more powerful.
Maybe instead of trying to retroactively fix the broken cp interface it would be better (and easier) to just allow cp to take a convenient list from the user...Then again doing aws ls my-prefix | grep 'REGEXP' | aws cp -T- . will always be more powerful.
Maybe instead of trying to retroactively fix the broken cp interface it would be better (and easier) to just allow cp to take a convenient list from the user...

Because it follows the unix philosophy more closely, of each tool doing 1 thing well, composed together to do alot.

If we add some kind of flag (i dunno maybe `--quiet` or something), that would allow ls to output just the file names, then we could do:

```
aws s3 ls --recursive --quiet s3://aidand-misc/te | aws cp -T- .
```

This would give so much flexibility, and also lead to other opportunities like:

```
aws s3 ls --recursive --quiet s3://aidand-misc/te | aws rm -T-
```

We would just need to investigate the technical barriers in such an approach. With how it's currently built, not sure it's easy to vary the positional args given the presence of certain flags - but would need to look more closely to find out.

And also we would need to consider the implications on Windows. Not too familiar with powershell/cmd but I think piping is a thing in powershell?

**Comment by REslim30** [comment]
Actually, @WesselAtWork, I see what you're getting at. `--auto-append-delimiter` as a way to retroactively reconcile inconsistent apis while retaining backwards compatibility, is a pretty good solution.

The point I was trying to make was that `s3 ls --recursive` is weird behaviour and we shouldn't make `s3 cp --recursive .` copy that behaviour.

At this point users would be used to how `s3 ls --recursive` behaves, so `s3 cp --recursive --auto-append-delimiter=false` would actually make sense (unless they've never used s3 ls --recursive before, in which case they'd probably end up finding this GitHub issue). This makes the `--recursive` analogy less meaningful across all the commands, but what can you do.

I think we should explore the piping solution, and only if it's not possible use the `--auto-append-delimiter` solution because the former provides some cool opportunities while the latter just patches an non-ideal situation.

**Comment by REslim30** [comment]
On windows you would need to do something like:

```
aws s3 ls myprefix > temp.txt; aws cp -T temp.txt outdir/
```

I tried searching for a powershell analogy to `aws s3 cp -T - outdir/` but couldn't find anything.

It's a bit more verbose, but it's very obvious what everything does.

**Comment by captainfalcon23** [comment]
I’ve been watching this issue for a while, and wanted to weigh in. @REslim30 a few times you’ve suggested using s3 ls and then piping that to s3 cp or some other process. I think the whole point of this issue is we shouldn’t have to use two different commands and it should be entirely possible to do this server side with a single command. 

So ideally, if I have a bucket with millions of files which haven’t been sorted into “directories” ie not separated by /, I should easily be able to do something like “aws s3 cp s3://mybucket/pre .”

if the cli is auto adding a trailing slash and the only reason why this is broken, then wouldn’t the solution proposed work and if a user wants to explicitly change the behaviour, set —auto-append-delimiter to false. And of course it would all be documented in the help, and not break any existing behaviour. 

If there is some concern about changing any behaviours of cp, why not introduce a new/improved command, like cp2? Which might offer completely different behaviour from cp, and not align with unix cp (not that I think it needs to align). 

Just my 2c.

**Comment by REslim30** [comment]
Thanks for the input @captainfalcon23

> I think the whole point of this issue is we shouldn’t have to use two different commands and it should be entirely possible to do this server side with a single command.

Not necessarily, I don't think. I think most people just want a convenient, and fast way to download a set of s3 objects by prefix. This obviously can be achieved by a single command, but I don't think that means it can't be two. Some people upvoted this workaround:

![image](https://github.com/aws/aws-cli/assets/43259657/89df04c7-812b-41b2-8828-e65268530929)

Which shows, people just want a good way to set a bunch of objects by prefix.

But I see what you're saying. Doing it in two commands is less simple from the user's perspective. Since they need to understand the `-T -` part of the command and how things are piped together. So it might be more geared towards the power user. To me, however, the `--auto-append-delimiter` solution doesn't feel as intuitive as `s3 ls | s3 cp -T - .` and the power and flexibility that something like `s3 cp -T - .` would provide would be really cool (e.g., the list of s3 objects could really come from anywhere, not just `aws s3 ls`, could be a static file or an in-house script, or other SDKs). but I might be in the minority in that; would like to see what everyone else says.

> If there is some concern about changing any behaviours of cp, why not introduce a new/improved command, like cp2?

This could work, `s3cmd` has a get command:

![image](https://github.com/aws/aws-cli/assets/43259657/34770c01-1c83-4545-8fe7-c6fbf1be5585)

But from the aws teams perspective, this is an entire command they'd have to support. So there's a maintenance cost to consider.

**Comment by captainfalcon23** [comment]
I will check out both options as workarounds :) It would also be nice if we could do an ls based on 2 wild cards e.g.
```
aws s3 ls s3://bucket/*middle*
```

But I doubt the API even supports something so complex :)

**Comment by diego-santacruz** [comment]
Hi! I would like to reinforce the request.

When the bucket has millions of objects in the same pseudo-directory prefix and you only want to download a few hundred files with a common prefix, the --exclude and --include flags do not do it, because the command will have to list all the millions of objects and before applying the filters. Compare that with `aws s3 ls` which works almost instantly because it can match the object prefixes.

Using `aws s3 ls` and piping the object list to `aws s3 cp` does work, but it is painstakingly slow. It is fine for a handful of files, but not hundreds or thousands.

Can you please add the proposed solution with a additional flag that allows the user to stop adding the trailing slash? That would preserve backwards compatibility, it should not be complex to do or maintain, and would be a tremendous improvement when you need to download things like CloudFront logs for specific time ranges from an S3 bucket (CloudFront puts all the logs under the same "pseudo-directory" prefix, with a timestamp in the object name and there are gazillions of files).

**Comment by gaellafond** [comment]
Same issue here, with `rm` instead of `cp`.

Sorry if that has been reported already, that a long thread. I didn't read it all.

My `rm` command was not working (using prefix), so I used `ls` to diagnose the issue. The thing is that `ls` doesn't work the same way. It interpret the prefix the way you would expect, so it does output all the files I want to delete. `rm` needs the `--recursive` flag, otherwise it interpret the path as a single file instead of a prefix. When using `--recursive`, aws cli ends up adding a `/` at the end of the prefix:

File: `awscli/customizations/s3/fileformat.py`
```
def s3_format(self, path, dir_op):
    if dir_op:
        if not path.endswith('/'):
            path += '/'
        return path, True
```

The docs is just wrong:
> --recursive (boolean) Command is performed on all files or objects under the specified directory **or prefix**.

https://docs.aws.amazon.com/cli/latest/reference/s3/rm.html
https://docs.aws.amazon.com/cli/latest/reference/s3/cp.html

That's wrong. It simply doesn't work with prefix.

I would expect `rm --recursive` and `cp --recursive` to work just like `ls`, so we can use `ls` to craft the right prefix before calling `rm` or `cp`.
